src/python/pants/option/config.py

```diff
@@ -230,10 +244,187 @@ def get_value(self, section: str, option: str) -> Optional[str]:
   def options(self, section: str) -> List[str]:
     return self.parser.options(section)
 
+  @property
   def defaults(self) -> Mapping[str, str]:
     return self.parser.defaults()
 
 
+_TomlPrimitve = Union[bool, int, float, str]
+_TomlValue = Union[_TomlPrimitve, List[_TomlPrimitve]]
+
+
+@dataclass(frozen=True)
+class _TomlValues(_ConfigValues):
+  values: Dict[str, Any]
+
+  @staticmethod
+  def _is_an_option(option_value: Union[_TomlValue, Dict]) -> bool:
+    """Determine if the value is actually an option belonging to that section.
+
+    A value that looks like an option might actually be a subscope, e.g. the option value
+    `java` belonging to the section `cache` could actually be the section `cache.java`, rather
+    than the option `--cache-java`.
+
+    We must also handle the special syntax of `my_list_option.add` and `my_list_option.remove`.
+    """
+    return (
+      not isinstance(option_value, dict) or "add" in option_value or "remove" in option_value
```

Comment:
I'm not super clear on operator precedence here. I think this does what you want (`not (a or b or c)`, as opposed to `(not a) or b or c`), but I'm a big fan of parentheses for clarity because I'm too lazy to learn operator precedence.
```suggestion
      not (isinstance(option_value, dict) or "add" in option_value or "remove" in option_value)
```